lua/pl/utils.lua

```diff
@@ -605,6 +607,78 @@ function utils.bind2 (fn,p)
     return function(x,...) return fn(x,p,...) end
 end
 
+
+
+
+--- Deprecation
+-- @section deprecation
+
+
+--- A deprecation warning function, to be overridden.
+-- An application can override this function to support proper output of
+-- deprecation warnings. The warnings can be generated from libraries or
+-- functions by calling `utils.raise_deprecation`. By default this function
+-- doesn't do anything.
+--
+-- Note: only applications should override this function, libraries should not.
+-- @string msg the message to display/log
+-- @string trace the traceback from where the deprecated element was invoked
+-- @usage
+-- function utils.deprecation_warning(msg, trace)
+--   io.stderr:write(msg .. "\n" .. trace .."\n")
+-- end
+function utils.deprecation_warning(msg, trace)
+  -- this does nothing by default
+end
+
+
+--- raises a deprecation warning.
+-- For options see the usage example below.
+--
+-- Note: the `opts.version_deprecated` field is the last version in which
+-- a feature or option was NOT YET deprecated! Because when writing the code it
+-- is quite often not known in what version the code will land. But the last
+-- released version is usually known.
+-- @param opts options table
+-- @see utils.deprecation_warning
+-- @usage
+-- function stringx.islower(str)
+--   deprecation_warning {
+--     source = "Penlight " .. utils._VERSION,                  -- optional
+--     message = "function 'islower' was renamed to 'is_lower'" -- required
+--     version_removed = "2.0.0",                               -- optional
+--     version_deprecated = "1.2.3",                            -- optional
+--   }
+--   return stringx.is_lower(str)
+-- end
+-- -- output: "[Penlight 1.9.2] function 'islower' was renamed to 'is_lower' (deprecated after 1.2.3, scheduled for removal in 2.0.0)"
+function utils.raise_deprecation(opts)
+  utils.assert_arg(1, opts, "table")
+  if type(opts.message) ~= "string" then
+    error("field 'message' of the options table must be a string", 2)
+  end
+  local trace = debug.traceback("", 2):match("[\n%s]*(.-)$")
```

Comment:
We might need to make accommodations for LuaJIT here which does not provide this functionality by default.

Reply:
what do you mean exactly? can you elaborate?

Reply:
I don't know. I just proved myself wrong trying to put together an MWE, I need to re-evaluate what I thought LuaJIT & OpenResty wouldn't do out of the box.

https://github.com/sile-typesetter/sile/issues/1013